### PR TITLE
Reduce memory usage during accuracy computation in SFT trainer

### PR DESCRIPTION
**Pull request by mikasenghaas**
> Should merge either this PR or #887 

Previously, we compute accuracy on each micro batch we not only materialize the `logits` but also the `probs` tensor, leading to unnecessary GPU memory usage which slightly decreases throughput.

This PR **changes the accuracy** computation such that we don't materialize the `probs` tensor in memory, reducing the peak GPU memory usage. However, compared to #887, we do not gain much in throughput because we still do the accuracy computation.

**Benchmarking command**

```bash
uv run sft @ configs/debug/sft.toml \
  --data.type fake \
  --data.length variable \
  --data.seq-len 4096 \
  --data.batch-size 4 \
  --data.micro-batch-size 1 \
  --bench
```

**Current**

Peak GPU mem: 26.20 GB

Screenshot 2025-09-11 at 1 24 17 PM

**This PR**

Peak GPU mem: 23.88 GB

Screenshot 2025-09-11 at 1 32 16 PM

**Comparison**

Screenshot 2025-09-11 at 1 32 38 PM

### CODE REVIEW

**Comment by mikasenghaas** [comment]
Closed in favor of #887
